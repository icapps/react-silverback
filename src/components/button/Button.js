import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '..';

const Button = props => (
  <button type="button" className={`btn ${props.className}`} onClick={props.handleClick} disabled={props.isDisabled || props.isPending}>
    {props.icon && <img src={props.icon} alt="+" />}
    {props.isPending ? <Spinner hasContainer={false} spinnerClassName={'button-spinner'} /> : <span>{props.text}</span>}
    {props.rightIcon && <img src={props.rightIcon} alt="+" />}
  </button>
);

Button.propTypes = {
  className: PropTypes.string,
  handleClick: PropTypes.func.isRequired,
  text: PropTypes.string.isRequired,
  icon: PropTypes.string,
  rightIcon: PropTypes.string,
  isDisabled: PropTypes.bool,
  isPending: PropTypes.bool,
};

Button.defaultProps = {
  className: '',
  icon: undefined,
  rightIcon: undefined,
  isDisabled: false,
  isPending: false,
};

export default Button;
