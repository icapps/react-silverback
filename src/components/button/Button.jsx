import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import './button.css';

const THEMES = [ 'primary', 'secondary', 'inverse' ];

const Button = (props) => {
  const renderClickPart = (cssClasses) => {
    if (props.link) {
      return (
        <Link className={cssClasses} to={props.link}>
          {props.children}
        </Link>
      );
    } else if (props.ahref) {
      return (
        <a href={props.ahref} className={cssClasses}>
          {props.children}
        </a>
      );
    }
    return (
      <button onClick={props.onClickCallback} className={cssClasses}>
        {props.children}
      </button>
    );
  };

  const theme = THEMES.includes(props.theme) ? props.theme : THEMES[0];
  return renderClickPart(`btn btn-${theme}`);
};

Button.propTypes = {
  onClickCallback: PropTypes.func,
  ahref: PropTypes.string,
  link: PropTypes.string,
  theme: PropTypes.string,
};

Button.defaultProps = {
  onClickCallback: () => {},
  ahref: '',
  link: '',
  theme: THEMES[0],
};

export default Button;
