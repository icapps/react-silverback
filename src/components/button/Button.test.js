import React from 'react';
import {shallow} from 'enzyme';
import Button from './Button';

describe('Button Component', () => {
  test('Render it as a button', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<Button onClickCallback={mockFn}>This is a Button</Button>);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('button').simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });

  it("should render a 'a' link", () => {
    const wrapper = shallow(<Button href='http://google.com'>This is a href link!</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a react router Link', () => {
    const wrapper = shallow(<Button link='/home'>This is a React Router Link!</Button>);
    expect(wrapper).toMatchSnapshot();
  });
});
