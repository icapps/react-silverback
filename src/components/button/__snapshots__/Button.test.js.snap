// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Render it as a button 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    ahref=""
    link=""
    onClickCallback={[Function]}
    theme="primary"
>
    This is a Button
</Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "This is a Button",
      "className": "btn btn-primary",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": "This is a Button",
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "This is a Button",
        "className": "btn btn-primary",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "This is a Button",
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Button Component should render a 'a' link 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    ahref=""
    href="http://google.com"
    link=""
    onClickCallback={[Function]}
    theme="primary"
>
    This is a href link!
</Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "This is a href link!",
      "className": "btn btn-primary",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": "This is a href link!",
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "This is a href link!",
        "className": "btn btn-primary",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "This is a href link!",
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Button Component should render a react router Link 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    ahref=""
    link="/home"
    onClickCallback={[Function]}
    theme="primary"
>
    This is a React Router Link!
</Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": "This is a React Router Link!",
      "className": "btn btn-primary",
      "replace": false,
      "to": "/home",
    },
    "ref": null,
    "rendered": "This is a React Router Link!",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "This is a React Router Link!",
        "className": "btn btn-primary",
        "replace": false,
        "to": "/home",
      },
      "ref": null,
      "rendered": "This is a React Router Link!",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
